### PR TITLE
support nested filter specification types

### PR DESCRIPTION
**Pull request by j8seangel**
## Launch Checklist

 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [x] Briefly describe the changes in this PR. --> Support nested expressions and include suggestions for every filter type
 - [ ] Document any changes to public APIs.
 - [ ] Manually test the debug page.
 - [x] Add an entry inside this element for inclusion in the `maplibre-gl-js` changelog: `<changelog></changelog>`.

### CODE REVIEW

**Comment by j8seangel** [comment]
Hi @HarelM First of all, big thanks for your effort!
I opened the PR by mistake because I wanted to test the changes in my fork first but once open I'll keep working on this one. 

> It would be great to let us know what are you solving.

We are moving from mapbox-gl and I found some definition errors on our existing expressions so I thought I could collaborate updating their types definition (thanks again for moving to typescript! 🙇 )

This was the error:
image

Writing expressions has been always difficult for me, so typing them are even harder but I think we can have a generic solution to allow writing any kind of expression. And that is why I added a list of every filter available in the [documentation](https://maplibre.org/maplibre-gl-js-docs/style-spec/expressions/#lookup)

> Note that you are editing a generated file. Please make sure to fix the code generation.

I thought the `src/style-spec/types.ts` is a source type definition file, could you explain then a bit more how that works ?

> Also it would be great if you can check the relevant tick boxes in this template.

Sure, let me accommodate and understand all and will keep you updated once they are ready

> If there's an issue you are solving with the PR, even better.

I haven't found any, but If that is the workflow I can create one explaining the error and point it here

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,7 +6,7 @@
 
 ### 🐞 Bug fixes
 
-- *...Add new stuff here...*
+- Fix nested expressions types definition [#757](https://github.com/maplibre/maplibre-gl-js/pull/757)
```
Comment (HarelM) [inline CHANGELOG.md:9]: Move this under 2.0 as this version wasn't released yet.

File: src/style-spec/feature_filter/convert.ts [hunk 2]
```diff
@@ -92,7 +92,7 @@ function _convertFilter(filter: FilterSpecification, expectedTypes: ExpectedType
         const children = (filter as any).slice(1).map(f => _convertFilter(f, expectedTypes));
         return children.length > 1 ? ['all'].concat(children) : [].concat(...children);
     } else if (op === 'none') {
-        return ['!', _convertFilter(['any'].concat(filter.slice(1) as any), {})];
+        return ['!', _convertFilter(['any'].concat(filter.slice(1) as any) as any, {})];
```
Comment (HarelM) [inline src/style-spec/feature_filter/convert.ts:95]: Better not to add more "as any", instead try reducing them as much as possible.

File: src/style-spec/types.ts [hunk 3]
```diff
@@ -9,16 +9,34 @@ export type ResolvedImageSpecification = string;
 
 export type PromoteIdSpecification = {[_: string]: string} | string;
 
+export type FilterSpecificationInputType = string | number | boolean;
```
Comment (HarelM) [inline src/style-spec/types.ts:12]: Check the top of this file, it says which duke was used to generate this.
Use the package.json to run the codegen script.

**Comment by HarelM** [comment]
Please also fix the failing tests.
Thanks!! :-)

File: src/style-spec/types.ts (hunk 3 above)
Comment (j8seangel) [reply src/style-spec/types.ts:12]: Ouch 🤦  I'll read more next time, updated and thanks!

File: src/style-spec/feature_filter/convert.ts (hunk 2 above)
Comment (j8seangel) [reply src/style-spec/feature_filter/convert.ts]: 👍

**Comment by j8seangel** [comment]
@HarelM all comments addressed, ready to review

**Review by HarelM — APPROVED** [review]

**Comment by HarelM** [comment]
Thanks!!!

**Pull request merged**